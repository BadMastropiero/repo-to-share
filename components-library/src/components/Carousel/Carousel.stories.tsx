import {ComponentMeta, ComponentStory} from '@storybook/react';
import React from 'react';
import {GenericProps} from '../Generic/Generic';
import {Carousel} from '.';
import {CarouselProps} from './Carousel.interface';

import {Blog, Service2} from '../Card/Card.stories';

export default {
  title: 'Components/Carousel',
  component: Carousel,
} as ComponentMeta<React.FC<typeof Carousel>>;

export const Base: ComponentStory<React.FC<GenericProps & CarouselProps>> = args => (
  <Carousel {...args}>
    <Service2 titleText="Shape your project" />
    <Service2 />
    <Service2 titleText="Shape your skills" />
    <Service2 titleText="Shape your body" />
    <Service2 titleText="Shape your self" />
  </Carousel>
);

export const Blogs: ComponentStory<React.FC<GenericProps & CarouselProps>> = () => (
  <Carousel variant="blogs" slidesWidth={380} showControls showPagination>
    <Blog />
    <Blog />
    <Blog />
    <Blog />
    <Blog />
    <Blog />
    <Blog />
  </Carousel>
);

export const Services: ComponentStory<React.FC<GenericProps & CarouselProps>> = () => (
  <Carousel variant="services" showPagination infinite autoplayDelay={3000}>
    <Service2 titleText="Shape your project" />
    <Service2 />
    <Service2 titleText="Shape your skills" />
    <Service2 titleText="Shape your self" />
  </Carousel>
);

Base.args = {
  disabled: false,
  variant: 'services',
};

Services.parameters = {
  design: {
    type: 'figma',
    url: 'https://www.figma.com/file/epHXucNHjke2rES69iOsaS/TetraImpacts?node-id=1119%3A4051',
  },
};

Blogs.parameters = {
  design: {
    type: 'figma',
    url: 'https://www.figma.com/file/epHXucNHjke2rES69iOsaS/TetraImpacts?node-id=1030%3A3827',
  },
};
