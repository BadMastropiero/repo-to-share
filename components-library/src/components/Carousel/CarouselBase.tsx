import React from 'react';
import ExternCarousel, {CarouselPluginTypes, slidesToShowPlugin} from '@brainhubeu/react-carousel';

// Styles
import {StyledCarousel} from './Carousel.styles';

// Interfaces
import {CarouselBaseProps} from './Carousel.interface';
import DeckCarousel from './DeckCarousel/DeckCarousel';

const mapVariantToOffset = {
  blogs: 60,
  services: -400,
};

const maVariantToSlidesToShowPluginOptions = {
  blogs: undefined,
  services: {
    numberOfSlides: 2,
  },
};

const CarouselBase: React.FC<CarouselBaseProps> = ({
  children,
  variant,
  disabled,
  infinite,
  className,
  slidesWidth,
  onSlideChange,
  slideIndex,
  ...rest
}) => {
  if (variant === 'services') {
    return (
      <DeckCarousel
        slideIndex={slideIndex}
        onSlideChange={onSlideChange}
        disabled={disabled}
        infinite={infinite}
        {...rest}
      >
        {children}
      </DeckCarousel>
    );
  }

  const plugins: (string | CarouselPluginTypes)[] = [
    {
      resolve: slidesToShowPlugin,
      options: variant ? maVariantToSlidesToShowPluginOptions[variant] : null,
    },
  ];

  if (infinite) plugins.push('infinite');

  let offset;
  if (variant) offset = mapVariantToOffset[variant];

  return (
    <StyledCarousel
      as={ExternCarousel}
      value={slideIndex}
      onChange={onSlideChange}
      className={`${variant}-carousel ${className || ''}`}
      plugins={plugins}
      draggable={!disabled}
      offset={offset}
      itemWidth={slidesWidth}
    >
      {children}
    </StyledCarousel>
  );
};

export default CarouselBase;
