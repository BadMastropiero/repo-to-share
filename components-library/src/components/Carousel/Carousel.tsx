import React, {useEffect, useState} from 'react';

// Styles
import {CarouselWrapper, CarouselControls} from './Carousel.styles';

// Interfaces
import {CarouselProps} from './Carousel.interface';
import {Button} from '../Button';
import {ArrowLeft, ArrowRight} from '../Icon';
import CarouselBase from './CarouselBase';
import Dots from './Dots/Dots';

const Carousel: React.FC<CarouselProps> = ({
  children,
  infinite,
  showControls,
  showPagination,
  disabled,
  paginationClickable,
  onSlideChange,
  variant,
  slideIndex,
  ...rest
}) => {
  const [currentSlide, setCurrentSlide] = useState(slideIndex || 0);
  const slidesAmmount = Array.isArray(children) ? children.length : 1;

  const nextSlide = () => {
    // setCurrentSlide( (currentSlide + 1) % slidesAmmount)
    setCurrentSlide(currentSlide + 1);
  };

  const previousSlide = () => {
    // setCurrentSlide( (currentSlide - 1) % slidesAmmount)
    setCurrentSlide(currentSlide - 1);
  };

  // Updating current slide based on prop
  useEffect(() => {
    if (slideIndex && slideIndex !== currentSlide) setCurrentSlide(slideIndex);
  }, [slideIndex]);

  return (
    <CarouselWrapper className="ti-carousel-wrapper">
      <CarouselBase
        slideIndex={currentSlide}
        onSlideChange={index => {
          setCurrentSlide(index);
          if (onSlideChange) onSlideChange(index);
        }}
        infinite={infinite}
        variant={variant}
        disabled={disabled}
        role="listbox"
        {...rest}
      >
        {children}
      </CarouselBase>
      {(showControls || showPagination) && (
        <CarouselControls>
          {showControls && (
            <>
              <Button
                presetSize="large"
                afterIcon={ArrowLeft}
                onClick={previousSlide}
                disabled={disabled || (!infinite && currentSlide === 0)}
              />
              <Button
                presetSize="large"
                afterIcon={ArrowRight}
                onClick={nextSlide}
                disabled={disabled || (!infinite && currentSlide === slidesAmmount - 1)}
              />
            </>
          )}
          {showPagination && (
            <Dots
              ammount={slidesAmmount}
              className="ti-carousel-dots"
              value={currentSlide}
              // style={{margin: variant === 'services' ? 'auto' : undefined}}
              style={{margin: !showControls ? 'auto' : undefined}}
              setValue={paginationClickable && !disabled ? setCurrentSlide : undefined}
              role="progressbar"
            />
          )}
        </CarouselControls>
      )}
    </CarouselWrapper>
  );
};

export default Carousel;
