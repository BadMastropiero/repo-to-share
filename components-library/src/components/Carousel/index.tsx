import Carousel from './Carousel';
import CarouselBase from './CarouselBase';
import DeckCarousel from './DeckCarousel/DeckCarousel';
import {CarouselWrapper, CarouselControls} from './Carousel.styles';
import Dots from './Dots/Dots';

export {Carousel, CarouselBase, DeckCarousel, CarouselWrapper, CarouselControls, Dots};

export default Carousel;
