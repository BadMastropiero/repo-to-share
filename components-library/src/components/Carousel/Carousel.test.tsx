import {render, screen} from '@testing-library/react';
import {composeStories} from '@storybook/testing-react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';
import * as stories from './Carousel.stories';

// Every component that is returned maps 1:1 with the stories, but they already contain all decorators from story level, meta level and global level!
const {Base} = composeStories(stories);

describe('Carousel tests', () => {
  test('should render', () => {
    render(<Base />);
    expect(screen.getByText('Shape your ideas')).toBeInTheDocument();
  });

  test('should match snapshot', () => {
    const tree = render(<Base />);
    expect(tree).toMatchSnapshot();
  });

  it('should trigger next slide', async () => {
    let index = 0;
    const setIndex = (number: number) => {
      index = number;
    };

    const tree = render(<Base slideIndex={index} onSlideChange={setIndex} />);
    const nextCard = await tree.findByText('Shape your ideas'); // If story changes content, this should be updated
    userEvent.click(nextCard);
    expect(index).toEqual(1);
  });
});
