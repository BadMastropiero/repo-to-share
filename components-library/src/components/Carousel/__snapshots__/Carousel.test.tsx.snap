// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel tests should match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-gswNZR fOVEgv ti-carousel-wrapper"
      >
        <div
          class="sc-ipEyDJ hMCvJX"
          role="listbox"
        >
          <div
            class="sc-csuSiG jxlpTE current"
            role="option"
          >
            <div
              class="sc-idXgbr sc-hHTYSt ffaAEa bfWWrs sc-kgTSHT QoigH"
              height="380px"
              style="display: flex; flex-direction: column; justify-content: center; align-items: center;"
              width="900px"
            >
              <div
                class="sc-hBxehG iiGtT"
              >
                <h1
                  class="sc-hLBbgP sc-iBYQkv gdgHTR"
                  font-weight="500"
                >
                  Shape your project
                </h1>
              </div>
              <div
                class="sc-ksBlkl fwTNuS"
              >
                <div
                  class="sc-hLBbgP sc-iBYQkv gbrESI"
                  style="max-width: 500px; text-align: center;"
                >
                  Consulting. Get the insights you need about your market and how to build your business with analysis and technology awareness.
                </div>
              </div>
            </div>
          </div>
          <div
            class="sc-csuSiG jxlpTE after"
            role="option"
          >
            <div
              class="sc-idXgbr sc-hHTYSt ffaAEa bfWWrs sc-kgTSHT QoigH"
              height="380px"
              style="display: flex; flex-direction: column; justify-content: center; align-items: center;"
              width="900px"
            >
              <div
                class="sc-hBxehG iiGtT"
              >
                <h1
                  class="sc-hLBbgP sc-iBYQkv gdgHTR"
                  font-weight="500"
                >
                  Shape your ideas
                </h1>
              </div>
              <div
                class="sc-ksBlkl fwTNuS"
              >
                <div
                  class="sc-hLBbgP sc-iBYQkv gbrESI"
                  style="max-width: 500px; text-align: center;"
                >
                  Consulting. Get the insights you need about your market and how to build your business with analysis and technology awareness.
                </div>
              </div>
            </div>
          </div>
          <div
            class="sc-csuSiG jxlpTE back"
            role="option"
          >
            <div
              class="sc-idXgbr sc-hHTYSt ffaAEa bfWWrs sc-kgTSHT QoigH"
              height="380px"
              style="display: flex; flex-direction: column; justify-content: center; align-items: center;"
              width="900px"
            >
              <div
                class="sc-hBxehG iiGtT"
              >
                <h1
                  class="sc-hLBbgP sc-iBYQkv gdgHTR"
                  font-weight="500"
                >
                  Shape your skills
                </h1>
              </div>
              <div
                class="sc-ksBlkl fwTNuS"
              >
                <div
                  class="sc-hLBbgP sc-iBYQkv gbrESI"
                  style="max-width: 500px; text-align: center;"
                >
                  Consulting. Get the insights you need about your market and how to build your business with analysis and technology awareness.
                </div>
              </div>
            </div>
          </div>
          <div
            class="sc-csuSiG jxlpTE back"
            role="option"
          >
            <div
              class="sc-idXgbr sc-hHTYSt ffaAEa bfWWrs sc-kgTSHT QoigH"
              height="380px"
              style="display: flex; flex-direction: column; justify-content: center; align-items: center;"
              width="900px"
            >
              <div
                class="sc-hBxehG iiGtT"
              >
                <h1
                  class="sc-hLBbgP sc-iBYQkv gdgHTR"
                  font-weight="500"
                >
                  Shape your body
                </h1>
              </div>
              <div
                class="sc-ksBlkl fwTNuS"
              >
                <div
                  class="sc-hLBbgP sc-iBYQkv gbrESI"
                  style="max-width: 500px; text-align: center;"
                >
                  Consulting. Get the insights you need about your market and how to build your business with analysis and technology awareness.
                </div>
              </div>
            </div>
          </div>
          <div
            class="sc-csuSiG jxlpTE back"
            role="option"
          >
            <div
              class="sc-idXgbr sc-hHTYSt ffaAEa bfWWrs sc-kgTSHT QoigH"
              height="380px"
              style="display: flex; flex-direction: column; justify-content: center; align-items: center;"
              width="900px"
            >
              <div
                class="sc-hBxehG iiGtT"
              >
                <h1
                  class="sc-hLBbgP sc-iBYQkv gdgHTR"
                  font-weight="500"
                >
                  Shape your self
                </h1>
              </div>
              <div
                class="sc-ksBlkl fwTNuS"
              >
                <div
                  class="sc-hLBbgP sc-iBYQkv gbrESI"
                  style="max-width: 500px; text-align: center;"
                >
                  Consulting. Get the insights you need about your market and how to build your business with analysis and technology awareness.
                </div>
              </div>
            </div>
          </div>
          <div
            class="sc-csuSiG jxlpTE layout-builder"
          >
            <div
              class="sc-idXgbr sc-hHTYSt ffaAEa bfWWrs sc-kgTSHT QoigH"
              height="380px"
              style="display: flex; flex-direction: column; justify-content: center; align-items: center;"
              width="900px"
            >
              <div
                class="sc-hBxehG iiGtT"
              >
                <h1
                  class="sc-hLBbgP sc-iBYQkv gdgHTR"
                  font-weight="500"
                >
                  Shape your project
                </h1>
              </div>
              <div
                class="sc-ksBlkl fwTNuS"
              >
                <div
                  class="sc-hLBbgP sc-iBYQkv gbrESI"
                  style="max-width: 500px; text-align: center;"
                >
                  Consulting. Get the insights you need about your market and how to build your business with analysis and technology awareness.
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-gswNZR fOVEgv ti-carousel-wrapper"
    >
      <div
        class="sc-ipEyDJ hMCvJX"
        role="listbox"
      >
        <div
          class="sc-csuSiG jxlpTE current"
          role="option"
        >
          <div
            class="sc-idXgbr sc-hHTYSt ffaAEa bfWWrs sc-kgTSHT QoigH"
            height="380px"
            style="display: flex; flex-direction: column; justify-content: center; align-items: center;"
            width="900px"
          >
            <div
              class="sc-hBxehG iiGtT"
            >
              <h1
                class="sc-hLBbgP sc-iBYQkv gdgHTR"
                font-weight="500"
              >
                Shape your project
              </h1>
            </div>
            <div
              class="sc-ksBlkl fwTNuS"
            >
              <div
                class="sc-hLBbgP sc-iBYQkv gbrESI"
                style="max-width: 500px; text-align: center;"
              >
                Consulting. Get the insights you need about your market and how to build your business with analysis and technology awareness.
              </div>
            </div>
          </div>
        </div>
        <div
          class="sc-csuSiG jxlpTE after"
          role="option"
        >
          <div
            class="sc-idXgbr sc-hHTYSt ffaAEa bfWWrs sc-kgTSHT QoigH"
            height="380px"
            style="display: flex; flex-direction: column; justify-content: center; align-items: center;"
            width="900px"
          >
            <div
              class="sc-hBxehG iiGtT"
            >
              <h1
                class="sc-hLBbgP sc-iBYQkv gdgHTR"
                font-weight="500"
              >
                Shape your ideas
              </h1>
            </div>
            <div
              class="sc-ksBlkl fwTNuS"
            >
              <div
                class="sc-hLBbgP sc-iBYQkv gbrESI"
                style="max-width: 500px; text-align: center;"
              >
                Consulting. Get the insights you need about your market and how to build your business with analysis and technology awareness.
              </div>
            </div>
          </div>
        </div>
        <div
          class="sc-csuSiG jxlpTE back"
          role="option"
        >
          <div
            class="sc-idXgbr sc-hHTYSt ffaAEa bfWWrs sc-kgTSHT QoigH"
            height="380px"
            style="display: flex; flex-direction: column; justify-content: center; align-items: center;"
            width="900px"
          >
            <div
              class="sc-hBxehG iiGtT"
            >
              <h1
                class="sc-hLBbgP sc-iBYQkv gdgHTR"
                font-weight="500"
              >
                Shape your skills
              </h1>
            </div>
            <div
              class="sc-ksBlkl fwTNuS"
            >
              <div
                class="sc-hLBbgP sc-iBYQkv gbrESI"
                style="max-width: 500px; text-align: center;"
              >
                Consulting. Get the insights you need about your market and how to build your business with analysis and technology awareness.
              </div>
            </div>
          </div>
        </div>
        <div
          class="sc-csuSiG jxlpTE back"
          role="option"
        >
          <div
            class="sc-idXgbr sc-hHTYSt ffaAEa bfWWrs sc-kgTSHT QoigH"
            height="380px"
            style="display: flex; flex-direction: column; justify-content: center; align-items: center;"
            width="900px"
          >
            <div
              class="sc-hBxehG iiGtT"
            >
              <h1
                class="sc-hLBbgP sc-iBYQkv gdgHTR"
                font-weight="500"
              >
                Shape your body
              </h1>
            </div>
            <div
              class="sc-ksBlkl fwTNuS"
            >
              <div
                class="sc-hLBbgP sc-iBYQkv gbrESI"
                style="max-width: 500px; text-align: center;"
              >
                Consulting. Get the insights you need about your market and how to build your business with analysis and technology awareness.
              </div>
            </div>
          </div>
        </div>
        <div
          class="sc-csuSiG jxlpTE back"
          role="option"
        >
          <div
            class="sc-idXgbr sc-hHTYSt ffaAEa bfWWrs sc-kgTSHT QoigH"
            height="380px"
            style="display: flex; flex-direction: column; justify-content: center; align-items: center;"
            width="900px"
          >
            <div
              class="sc-hBxehG iiGtT"
            >
              <h1
                class="sc-hLBbgP sc-iBYQkv gdgHTR"
                font-weight="500"
              >
                Shape your self
              </h1>
            </div>
            <div
              class="sc-ksBlkl fwTNuS"
            >
              <div
                class="sc-hLBbgP sc-iBYQkv gbrESI"
                style="max-width: 500px; text-align: center;"
              >
                Consulting. Get the insights you need about your market and how to build your business with analysis and technology awareness.
              </div>
            </div>
          </div>
        </div>
        <div
          class="sc-csuSiG jxlpTE layout-builder"
        >
          <div
            class="sc-idXgbr sc-hHTYSt ffaAEa bfWWrs sc-kgTSHT QoigH"
            height="380px"
            style="display: flex; flex-direction: column; justify-content: center; align-items: center;"
            width="900px"
          >
            <div
              class="sc-hBxehG iiGtT"
            >
              <h1
                class="sc-hLBbgP sc-iBYQkv gdgHTR"
                font-weight="500"
              >
                Shape your project
              </h1>
            </div>
            <div
              class="sc-ksBlkl fwTNuS"
            >
              <div
                class="sc-hLBbgP sc-iBYQkv gbrESI"
                style="max-width: 500px; text-align: center;"
              >
                Consulting. Get the insights you need about your market and how to build your business with analysis and technology awareness.
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
